src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1727 @@
+   */
+  protected boolean useCatalog() {
+    return false;
+  }
+
+  @Override
+  public List<TableId> tableIds(Connection conn) throws SQLException {
+    DatabaseMetaData metadata = conn.getMetaData();
+    String[] tableTypes = tableTypes(metadata, this.tableTypes);
+
+    try (ResultSet rs = metadata.getTables(catalogPattern(), schemaPattern(), "%", tableTypes)) {
+      List<TableId> tableIds = new ArrayList<>();
+      while (rs.next()) {
+        String catalogName = rs.getString(1);
+        String schemaName = rs.getString(2);
+        String tableName = rs.getString(3);
+        TableId tableId = new TableId(catalogName, schemaName, tableName);
+        if (includeTable(tableId)) {
+          tableIds.add(tableId);
+        }
+      }
+      return tableIds;
+    }
+  }
+
+  protected String catalogPattern() {
+    return catalogPattern;
+  }
+
+  protected String schemaPattern() {
+    return schemaPattern;
+  }
+
+  /**
+   * Determine whether the table with the specific name is to be included in the tables.
+   *
+   * <p>This method can be overridden to exclude certain database tables.
+   *
+   * @param table the identifier of the table; may be null
+   */
+  protected boolean includeTable(TableId table) {
+    return true;
+  }
+
+  /**
+   * Find the available table types that are returned by the JDBC driver that case insensitively
+   * match the specified types.
+   *
+   * @param metadata the database metadata; may not be null but may be empty if no table types
+   * @param types    the case-independent table types that are desired
+   * @return the array of table types take directly from the list of available types returned by the
+   *     JDBC driver; never null
+   * @throws SQLException if there is an error with the database connection
+   */
+  protected String[] tableTypes(
+      DatabaseMetaData metadata,
+      Set<String> types
+  ) throws SQLException {
+    // Compute the uppercase form of the desired types ...
+    Set<String> uppercaseTypes = new HashSet<>();
+    for (String type : types) {
+      if (type != null) {
+        uppercaseTypes.add(type.toUpperCase());
+      }
+    }
+    // Now find out the available table types ...
+    Set<String> matchingTableTypes = new HashSet<>();
+    try (ResultSet rs = metadata.getTableTypes()) {
+      while (rs.next()) {
+        String tableType = rs.getString(1);
+        if (tableType != null && uppercaseTypes.contains(tableType.toUpperCase())) {
+          matchingTableTypes.add(tableType);
+        }
+      }
+    }
+    return matchingTableTypes.toArray(new String[matchingTableTypes.size()]);
+  }
+
+  @Override
+  public IdentifierRules identifierRules() {
+    if (identifierRules.get() == null) {
+      if (defaultIdentifierRules != null && connections.peek() == null) {
+        // We've not yet established a connection to the database, so in this case
+        // just return the default rules ...
+        return defaultIdentifierRules;
```

Comment:
I was trying to avoid creating a connection just to be able to use the identifier rules in the generic case. But I think you're right -- that behavior is unexpected. All of the concrete types define specific identifier rules, so it's only the generic case that needs to talk to the database. Later in the method, If we can't establish a connection, we log a warning and use the defaults. So, I'll remove this if-block altogether.